File: src/test/java/com/oz/office_tastezip/support/AuthTestHelper.java
package com.oz.office_tastezip.support;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.oz.office_tastezip.domain.auth.dto.LoginDto;
import com.oz.office_tastezip.global.util.RSAUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Objects;

@Component
public class AuthTestHelper {

    private TestRestTemplate restTemplate;
    @Autowired
    private ObjectMapper objectMapper;

    private String host;
    private int port;
    private String publicKey;

    public void setHostAndPort(String host, int port, String publicKey, TestRestTemplate restTemplate) {
        this.host = host;
        this.port = port;
        this.publicKey = publicKey;
        this.restTemplate = restTemplate;
    }

    public AuthTokens loginAndGetTokens(String email, String password) throws Exception {
        LoginDto loginDto = new LoginDto(email, RSAUtils.encrypt(password, publicKey));
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<LoginDto> request = new HttpEntity<>(loginDto, headers);

        ResponseEntity<String> loginRes = restTemplate.postForEntity(host + port + "/api/v1/otz/auth/login", request, String.class);

        String accessToken = objectMapper.readTree(loginRes.getBody()).get("data").get("accessToken").asText();
        String refreshToken = extractRefreshTokenFromSetCookie(Objects.requireNonNull(loginRes.getHeaders().get("Set-Cookie")));

        return new AuthTokens(accessToken, refreshToken);
    }

    private String extractRefreshTokenFromSetCookie(List<String> setCookies) {
        return setCookies.stream()
                .filter(cookie -> cookie.startsWith("refreshToken="))
                .map(cookie -> cookie.split("=")[1].split(";")[0])
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("refreshToken 쿠키 없음"));
    }

}

File: src/test/kotlin/com/oz/office_tastezip/support/AuthTestHelper.kt
package com.oz.office_tastezip.support

import com.fasterxml.jackson.databind.ObjectMapper
import com.oz.office_tastezip.domain.auth.dto.LoginDto
import com.oz.office_tastezip.global.util.RSAUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.stereotype.Component

@Component
class AuthTestHelper @Autowired constructor(
    private val objectMapper: ObjectMapper
) {

    private lateinit var restTemplate: TestRestTemplate
    private lateinit var host: String
    private var port: Int = 0
    private lateinit var publicKey: String

    fun setHostAndPort(host: String, port: Int, publicKey: String, restTemplate: TestRestTemplate) {
        this.host = host
        this.port = port
        this.publicKey = publicKey
        this.restTemplate = restTemplate
    }

    fun loginAndGetTokens(email: String, password: String): AuthToken {
        val loginDto = LoginDto(email, RSAUtils.encrypt(password, publicKey))

        val headers = HttpHeaders().apply {
            contentType = MediaType.APPLICATION_JSON
        }

        val request = HttpEntity(loginDto, headers)
        val loginRes = restTemplate.postForEntity(
            "$host$port/api/v1/otz/auth/login",
            request,
            String::class.java
        )

        val responseJson = objectMapper.readTree(loginRes.body)
        val accessToken = responseJson["data"]["accessToken"].asText()
        val refreshToken = extractRefreshTokenFromSetCookie(
            loginRes.headers["Set-Cookie"]
                ?: throw IllegalStateException("Set-Cookie 헤더 없음")
        )

        return AuthToken(accessToken, refreshToken)
    }

    private fun extractRefreshTokenFromSetCookie(setCookies: List<String>): String {
        return setCookies.firstOrNull { it.startsWith("refreshToken=") }
            ?.substringAfter("refreshToken=")
            ?.substringBefore(";")
            ?: throw IllegalStateException("refreshToken 쿠키 없음")
    }
}
